zerver/lib/actions.py

```diff
@@ -1737,6 +1737,8 @@ def do_add_reaction(user_profile: UserProfile, message: Message,
     reaction = Reaction(user_profile=user_profile, message=message,
                         emoji_name=emoji_name, emoji_code=emoji_code,
                         reaction_type=reaction_type)
+    message.has_reaction = True
+    message.save(update_fields=["has_reaction"])
```

Comment:
We should check if `has_reaction` was already True, and only do the write if it wasn't.